src/analyses/variable-sensitivity/dependency_context_abstract_object.h

```diff
@@ -0,0 +1,199 @@
+   * abstract_object with a given set of locations.
+   */
+  class location_update_visitort:
+    public abstract_objectt::abstract_object_visitort
+  {
+  public:
+    explicit location_update_visitort(const locationst &locations):
+      locations(locations) { }
+
+    abstract_object_pointert visit(const abstract_object_pointert element) const
+    {
+      return element->update_location_context(locations, true);
+    }
+
+    locationst locations;
+  };
+
+  locationst get_location_union(const locationst &locations) const;
+
+  virtual const typet &type() const
+  {
+    return child_abstract_object->type();
+  }
+
+  virtual bool is_top() const override
+  {
+    return child_abstract_object->is_top();
+  }
+
+  virtual bool is_bottom() const override
+  {
+    return child_abstract_object->is_bottom();
+  }
+
+  abstract_object_pointert expression_transform(
+    const exprt &expr,
+    const abstract_environmentt &environment,
+    const namespacet &ns) const;
+
+  virtual exprt to_constant() const override
+  {
+    return child_abstract_object->to_constant();
+  }
+
+  virtual void output(
+    std::ostream &out, const class ai_baset &ai, const namespacet &ns) const
+  override;
+
+
+protected:
+  CLONE
+
+  virtual abstract_object_pointert merge(
+    abstract_object_pointert other) const override;
+
+  virtual abstract_object_pointert read(
+    const abstract_environmentt &env,
+    const exprt &specifier,
+    const namespacet &ns) const override;
+
+  virtual abstract_object_pointert write(
+    abstract_environmentt &environment,
+    const namespacet &ns,
+    const std::stack<exprt> stack,
+    const exprt &specifier,
+    const abstract_object_pointert value,
+    bool merging_write) const override;
+
+  static void output_last_written_locations(
+    std::ostream &out,
+    const abstract_objectt::locationst &locations);
+
+  virtual abstract_objectt::locationst get_last_written_locations() const;
+
+private:
+  // To enforce copy-on-write these are private and have read-only accessors
+  abstract_objectt::locationst last_written_locations;
+  abstract_object_pointert child_abstract_object;
+
+  // These are internal hooks that allow sub-classes to perform additional
+  // actions when an abstract_object is set/unset to TOP
+  virtual void make_top_internal() override;
+  virtual void clear_top_internal() override;
+
+  virtual abstract_object_pointert abstract_object_merge_internal(
+    const abstract_object_pointert other) const override;
+
+  void set_last_written_locations(
+    const abstract_objectt::locationst &locations);
+
+  void set_child(
+    const abstract_object_pointert &child);
+};
+
+
+
+/// Templated extension of the abstract implementation, used as a wrapper around
+/// other abstract_objectt classes to enable the factory to instantiate the
+/// context information
+template <class AOT>
+class context_abstract_objectt: public dependency_context_abstract_objectt
+{
```

Comment:
This makes me mildly nervous but I can accept an answer of "no really, this *is* the easy way".

Reply:
:-) In general, doing it this way meant that the abstract_object_factory only had to have very minimal changes. When we come to refactor that as well, then probably there are better ways to do this., but at the moment I'm going to claim 'this is the easy way' unless there is something more specific than just the templating that is making you worried?